src/realm/object-store/impl/object_notifier.cpp

```diff
@@ -32,33 +32,86 @@ ObjectNotifier::ObjectNotifier(std::shared_ptr<Realm> realm, TableKey table, Obj
 
 bool ObjectNotifier::do_add_required_change_info(TransactionChangeInfo& info)
 {
+    // The object might have been deleted in the meantime.
+    if (!m_table)
+        return false;
+
     m_info = &info;
-    info.tables[m_table.value];
-    return false;
+    info.tables[m_table->get_key().value];
+
+    // When adding or removing a callback the related tables can change due to the way we calculate related tables
+    // when key path filters are set hence we need to recalculate every time the callbacks are changed.
+    util::CheckedLockGuard lock(m_callback_mutex);
+    if (m_did_modify_callbacks) {
+        m_related_tables = {};
+        recalculate_key_path_arrays();
+        DeepChangeChecker::find_filtered_related_tables(m_related_tables, *m_table, m_key_path_arrays);
+        m_did_modify_callbacks = false;
+    }
+
+    return true;
 }
 
 void ObjectNotifier::run()
 {
     if (!m_table)
         return;
 
-    auto it = m_info->tables.find(m_table.value);
+    if (!m_change.modifications.contains(0) && any_callbacks_filtered()) {
```

Comment:
I wasn't involved in the original implementation but from what I understood:
We do add the index of modified objects for Results, Lists, etc.
In case of a single object we just mark the object itself as modified (where the properties are then shown in `m_change.columns`).
I guess the reason for this is that `ObjectNotifier` is a `CollectionNotifier` but not everything works the same, this being one of those things.

Reply:
Yeah, it's a hack to work around this code being designed for collections.